src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                AddEntries2(preference, isUnusedLocalAssignment: false);
+            }
+
+            void AddEntries2(UnusedValuePreference preference, bool isUnusedLocalAssignment)
+            {
+                AddEntryCore(preference, isUnusedLocalAssignment, isRemovableAssignment: true);
+                AddEntryCore(preference, isUnusedLocalAssignment, isRemovableAssignment: false);
+            }
+
+            void AddEntryCore(UnusedValuePreference preference, bool isUnusedLocalAssignment, bool isRemovableAssignment)
+            {
+                var propertiesBuilder = ImmutableDictionary.CreateBuilder<string, string>();
+
+                propertiesBuilder.Add(UnusedValuePreferenceKey, preference.ToString());
+                if (isUnusedLocalAssignment)
+                {
+                    propertiesBuilder.Add(IsUnusedLocalAssignmentKey, string.Empty);
+                }
+                if (isRemovableAssignment)
+                {
+                    propertiesBuilder.Add(IsRemovableAssignmentKey, string.Empty);
+                }
+
+                builder.Add((preference, isUnusedLocalAssignment, isRemovableAssignment), propertiesBuilder.ToImmutable());
+            }
+        }
+
+        protected abstract Location GetDefinitionLocationToFade(IOperation unusedDefinition);
+        protected abstract bool SupportsDiscard(SyntaxTree tree);
+        protected abstract Option<CodeStyleOption<UnusedValuePreference>> UnusedValueExpressionStatementOption { get; }
+        protected abstract Option<CodeStyleOption<UnusedValuePreference>> UnusedValueAssignmentOption { get; }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        // Our analysis is limited to unused expressions in a code block, hence is unaffected by changes outside the code block.
+        // Hence, we can support incremental span based method body analysis.
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(
+                compilationContext => SymbolStartAnalyzer.CreateAndRegisterActions(
+                    compilationContext, GetDefinitionLocationToFade, SupportsDiscard,
+                    UnusedValueExpressionStatementOption, UnusedValueAssignmentOption));
+
+        private sealed class SymbolStartAnalyzer
+        {
+            private readonly Func<IOperation, Location> _getDefinitionLocationToFade;
+            private readonly Func<SyntaxTree, bool> _supportsDiscard;
+            private readonly Option<CodeStyleOption<UnusedValuePreference>> _unusedValueExpressionStatementOption;
+            private readonly Option<CodeStyleOption<UnusedValuePreference>> _unusedValueAssignmentOption;
+
+            private readonly INamedTypeSymbol _eventArgsType, _interlockedType, _immutableInterlockedType;
```

Comment:
presumably these types may not exist in the compilation. as such they may be 'null' right?  If so, can they be named 'Opt' so it's clear that htey need to be handled with care?